Kubernetes: Refresh ServiceAccount token

linki opened the pull request:
This PR adds support for expiring service account tokens by refreshing them from their mount path regularly.

Fixes https://github.com/zalando/skipper/issues/1277 and https://github.com/zalando/skipper/issues/1283

The implementation is inspired by [Kubernetes way of handling it](https://github.com/kubernetes/kubernetes/blob/v1.17.2/staging/src/k8s.io/client-go/transport/token_source.go#L48-L61).

### How it works

It changes `clusterClient` to use a [`oauth2.TokenSource`](https://godoc.org/golang.org/x/oauth2#TokenSource) instead of a plain string `token`. It then uses a custom `fileTokenSource` that fetches the token from a file, basically rebuilding the current functionality with the exception that it reads the file every single time. Finally, it wraps the file-based TokenSoure into a [`oauth2.ReuseTokenSource`](https://godoc.org/golang.org/x/oauth2#ReuseTokenSource) which takes care of reading the token from cache and only reloading it once it's expired.

Interesting to note here is that the actual expiration time from the token read from the file is ignored and statically set to 1 minute instead, resulting in re-reading the token once per minute. This is consistent with `client-go`'s implementation but doesn't have to be good since `client-go` assumes a token lifetime of 10 minutes in future Kubernetes clusters.

I assume periodically re-reading regardless of the token's expiration time is an attempt to work well in case tokens get revoked (by deleting the ServiceAccount). A local client that only has the token isn't able to detect that case. Of cource, one can try reading a new token if there's an unexpected `401` but that would complicate the code further.

The current implementation strikes a balance between simple and correct. The tokens are mounted as `tmpfs` so the overhead of re-reading should be rather small.

Let me know what you think.

@szuecs 
@aryszka

szuecs commented:
@linki code looks good but check travis output please

vet: dataclients/kubernetes/kube_test.go:2355:15: undeclared name: readServiceAccountToken

linki commented:
Thanks, I missed that.

I changed the test to cover the missing file case as before but I removed the case where it's "out-of-cluster" because that's not handled by the TokenSource anymore.

szuecs commented:
One question would be why not using the internal secrets module, but use a dependency from x?
We can also later change it to our net.Client, that is in development.

aryszka commented on dataclients/kubernetes/token_source.go:
```diff
@@ -0,0 +1,26 @@
+
+func (ts *fileTokenSource) Token() (*oauth2.Token, error) {
+	token, err := ioutil.ReadFile(ts.path)
+	if err != nil {
```
in case of IO errors, would it make sense to return the previous token? Or is this handled by oauth2.ReuseTokenSource?

aryszka replied:
discussed IRL, it's fine

aryszka commented:
generally lgtm. See my question about the IO errors.

I also wonder if it would make sense to make this a part of the "secrets" package, and use it with the net.Client implementation. We can also do that in a separate PR, too.

aryszka commented:
:+1:

linki commented:
Changed implementation to use skipper's `secrets` package.

szuecs commented:
:+1:

aryszka commented:
:+1:
